spec.bs

```diff
@@ -0,0 +1,369 @@
+  1. Let |epoch-switch-time-decision-duration| = Seconds(|epoch-switch-time-decision-hash| % 604800) (i.e. 604800 is 7 days in seconds).
+  1. Let |now| be the current time.
+  1. If |now| - |epoch-switch-time-decision-duration| > |last-epoch-time|:
+      1. Let |result| = the last up to 3 epochs from [=user-topics-state=].epochs (i.e. [=user-topics-state=].epochs[-1], [-2], and [-3] for any valid index).
+  1. Else:
+      1. Let |result| = the last up to 3 epochs from [=user-topics-state=].epochs except for the last one (i.e. [=user-topics-state=].epochs[-2], [-3], and [-4] for any valid index).
+  1. Return |result|.
+
+  <h2 id="topics-for-caller-header">Topics for caller</h2>
+  For a [=topics-caller-context=] |C|, the <dfn>topics for caller</dfn> method steps are:
+  1. Let |epochs| be the result of running [=epochs for caller=] given |C| as input.
+  1. Let |result| be an empty list [].
+  1. For each |epoch| in |epochs|:
+    1. Let |topic| be |undefined|.
+    1. Let |random-or-top-topic-decision-hash| = 64 bit truncation of HMAC-SHA256("random-or-top-topic-decision", [=user-topics-state=].hmac-key, |epoch|.id, |C|.page.top-level-domain).
+    1. If |random-or-top-topic-decision-hash| % 100 < 5:
+      1. Let |random-topic-index-decision-hash| = 64 bit truncation of  HMAC-SHA256("random-topic-index-decision", [=user-topics-state=].hmac-key, |epoch|.id, |C|.page.top-level-domain).
+      1. Let |random-topic-index| = |random-topic-index-decision-hash| % Length(|epoch|.taxonomy)
+      1. Set |topic| = {{BrowsingTopic}} {topic: |epoch|.taxonomy[|random-topic-index|], configVersion: epoch.config-version, modelVersion: epoch.model-version, taxonomyVersion: epoch.taxonomy-version, version = <a href="#version">combined version</a>}.
+    1. Else:
+        1. Let |top-topic-index-decision-hash| = 64 bit truncation of HMAC-SHA256("top-topic-index-decision", [=user-topics-state=].hmac-key, |epoch|.id, |C|.page.top-level-domain).
+        1. Let |top-topic-index| = |top-topic-index-decision-hash| % 5.
+        1. If |epoch|.top-5-topics-with-caller-origins[|top-topic-index|].caller-origins contains |C|.caller-origin:
+            1. Set |topic| = {{BrowsingTopic}} {topic: |epoch|.top-5-topics-with-caller-origins[|top-topic-index|].topic, configVersion: configVersion: epoch.config-version, modelVersion: epoch.model-version, taxonomyVersion: epoch.taxonomy-version, version = <a href="#version">combined version</a>}.
+    1. If |topic| is not |undefined|, add |topic| to |result|.
+  1. Remove duplicate entries in |result|.
+  1. Random shuffle entries in |result|.
+  1. Return |result|.
+
+</section>
+
+<section>
+  <h2 id="the-javascript-api-header">The JavaScript API</h2>
+
+  The Topics API lives under the {{Document}} interface, and is only available if the document is in [=secure context=].
+
+  <pre class="idl">
+    dictionary BrowsingTopicsOptions {
+      boolean observe = true;
```

Comment:
Booleans must not default to false in web specs: https://webidl.spec.whatwg.org/#:~:text=It%20is%20strongly%20suggested%20not%20to%20use%20a%20default%20value%20of%20true%20for%20boolean%2Dtyped%20arguments